crates/audio/src/spatial.rs

```diff
@@ -0,0 +1,125 @@
+use bevy::prelude::*;
+use bevy_kira_audio::{
+    prelude::{Audio as KAudio, AudioSource as KAudioSource},
+    AudioControl, AudioInstance,
+};
+use de_camera::CameraFocus;
+use de_core::gamestate::GameState;
+
+pub(crate) struct SpatialSoundPlugin;
+
+impl Plugin for SpatialSoundPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_system(start.run_if(in_state(GameState::Playing)))
+            .add_system(
+                update_spatial
+                    .in_base_set(CoreSet::PostUpdate)
+                    .run_if(in_state(GameState::Playing)),
+            );
+    }
+}
+
+#[derive(Component, Default)]
+pub struct SpatialSound;
+
+#[derive(Component)]
+pub struct Volume(pub f32);
+
+impl Default for Volume {
+    fn default() -> Self {
+        Self(1.)
+    }
+}
+
+trait VolumeOptionExt {
+    fn volume(&self) -> f32;
+}
+
+impl VolumeOptionExt for Option<&Volume> {
+    fn volume(&self) -> f32 {
+        match self {
+            Some(volume) => volume.0,
+            None => 1.,
+        }
+    }
+}
+
+#[derive(Bundle, Default)]
+pub struct SpatialSoundBundle {
+    pub sound: Handle<KAudioSource>,
+    pub spatial_sound: SpatialSound,
+    pub volume: Volume,
+}
+
+fn calculate_volume_and_pan(
+    camera: &GlobalTransform,
+    focus: &CameraFocus,
+    sound: &GlobalTransform,
+) -> (f64, f64) {
+    let cam_right = camera.right();
+    let sound_dir = (sound.translation() - camera.translation()).normalize();
+    let pan = cam_right.dot(sound_dir) * 0.5 + 0.5;
+
+    let distance_from_camera = camera.translation().distance(sound.translation());
+    let camera_zoom_distance = focus.distance().inner();
+    let distance_attenuation =
```

Comment:
I would try to apply inverse-square law first and only make it linear if it doesn't sound good.

```rust
let camera_distance_squared = camera.translation().distance_squared(sound.translation());
// Anything closer than 70% of zoom distance is at full volume.
let base_distance_squared = (0.7 * focus.distance().inner()).powi(2);
// Inverse-square law
let distance_attenuation = (base_distance_squared / camera_distance_squared).min(1.);
```

Reply:
This is actually something I'd like to really look deeper into perhaps at a later time. I *feel* (although haven't experimented much, just tried to figure out what I want) like standard inverse-square/other distance-based attenuation might not be the best fit here.

Take an example scenario, I'm looking at a building with the furthest zoom level at a 45 degree angle from the ground. I obviously want to hear the sounds emitted by that building. But if there's another building the same distance away, directly behind the camera, I don't think it feels natural to play sounds from it, at least at the same level.

Not that this current implementation does anything in favor of that, just want to bring it up. I thought maybe I could get away with the sound overhaul in one PR and then tune the precise parameters, but I can experiment here as well ofc.

Reply:
I unfortunately don't have any other games quickly on hand, but I did some experimenting in LoL since it uses an RTS-like camera (and it has its deepest roots in RTS). In that game, sound volume seems to depend on distance from the camera focus (the area where the sound is audible seems like a circular area when clicking the minimap), with a short falloff near the edge of the audible range.

With that in mind, I've updated the PR with a proposal for a similar falloff function. I think something like it is a great fit for an RTS, although in this state it is certainly not perfect.

Reply:
> Take an example scenario, I'm looking at a building with the furthest zoom level at a 45 degree angle from the ground. I obviously want to hear the sounds emitted by that building. But if there's another building the same distance away, directly behind the camera, I don't think it feels natural to play sounds from it, at least at the same level.

I agree but this is IMHO orthogonal to the distance square law. Instead, we should either discard all sounds coming from outside the camera frustum or have another multiplicative attenuation factor based on direction. I would go for the first version to avoid over-engineering.

> Not that this current implementation does anything in favor of that, just want to bring it up. I thought maybe I could get away with the sound overhaul in one PR and then tune the precise parameters, but I can experiment here as well ofc.

:+1: Yep. I would not deal with the direction issue here.

The distance attenuation experiment should be easy (few lines of code) so I would try that. Lets do a few quick experiment, pick the best easy solution and create an issue for further experimentation / research on the topic.

> I unfortunately don't have any other games quickly on hand, but I did some experimenting in LoL since it uses an RTS-like camera (and it has its deepest roots in RTS). In that game, sound volume seems to depend on distance from the camera focus (the area where the sound is audible seems like a circular area when clicking the minimap), with a short falloff near the edge of the audible range.

Interesting. Can you push commented out versions so that I can try them all myself? :pray:

I was imagining that e.g. a drone flying high up (close to the camera) being louder might create a cool effect.

I think that a combination of the distance attenuation (from camera) + directional attenuation (angle from forward direction) might give nice results. I.e. close camera flybys would be loud while things near the center of the screen would be more dominant.

Reply:
Also, in the future, we should consider something like Reinhard mapping so that we don't have a hard clip at value 1.

Reply:
I have added a few options for both distance attenuation and view occlusion, accessible with debug controls that work like two sets of radio buttons.

**Distance attenuation**
1 = No attenuation
2 = Linear from focus
3 = Inverse square
4 = Inverse square from focus

**View occlusion**
8 = None
9 = Don't play sounds outside view
0 = Volume is multiplied by 1...0 from 22.5...45 degrees from center of view, calculated from a point a bit behind the camera so that sounds just behind the camera are also heard

Current personal favorite combo is inverse square attenuation + angle occlusion (3 + 0).